### PR TITLE
Add suggested maps section to plugin import screen

### PR DESCRIPTION
**Pull request by nnngrach**
https://github.com/osmandapp/OsmAnd-Issues/issues/639

Suggested maps video:
https://www.dropbox.com/s/1p13adp7t7tcfti/Screen%20Recording%202021-07-06%20at%2009.26.49.mov?dl=0

Suggested profiles video:
https://www.dropbox.com/s/etyp7n358ca96rc/Screen%20Recording%202021-07-08%20at%2020.08.23.mov?dl=0

### CODE REVIEW

File: Sources/Helpers/OAResourcesUIHelper.mm [hunk 1]
```diff
@@ -668,6 +845,11 @@ + (void) startDownloadOfCustomItem:(OACustomResourceItem *)item onTaskCreated:(O
 }
 
 + (void) offerDownloadAndInstallOf:(OARepositoryResourceItem *)item onTaskCreated:(OADownloadTaskCallback)onTaskCreated onTaskResumed:(OADownloadTaskCallback)onTaskResumed
+{
```
Comment (PaulStets) [inline Sources/Helpers/OAResourcesUIHelper.mm:848]: Do not pass view controllers to methods, instead create a method with a callback that will show the alert.

File: Sources/Controllers/2.0/Settings/ImportExport/OAPluginInstalledViewController.mm [hunk 2]
```diff
@@ -267,6 +294,36 @@ - (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForR
...
+        OAIconTextDescSwitchCell* cell = [tableView dequeueReusableCellWithIdentifier:[OAIconTextDescSwitchCell getCellIdentifier]];
+        if (cell == nil)
+        {
+            NSArray *nib = [[NSBundle mainBundle] loadNibNamed:[OAIconTextDescSwitchCell getCellIdentifier] owner:self options:nil];
+            cell = (OAIconTextDescSwitchCell *)[nib objectAtIndex:0];
+        }
+        OAApplicationMode *am = item[@"mode"];
+        BOOL isEnabled = [OAApplicationMode.values containsObject:am];
+        cell.separatorInset = UIEdgeInsetsMake(0.0, indexPath.row < OAApplicationMode.allPossibleValues.count - 1 ? 62.0 : 0.0, 0.0, 0.0);
+        cell.selectionStyle = UITableViewCellSelectionStyleNone;
```
Comment (PaulStets) [inline Sources/Controllers/2.0/Settings/ImportExport/OAPluginInstalledViewController.mm:309]: This line should only be called during cell initialization

File: Sources/Controllers/2.0/Settings/ImportExport/OAPluginInstalledViewController.mm [hunk 3]
```diff
@@ -303,21 +360,30 @@ - (UIView *) getHeaderForTableView:(UITableView *)tableView withFirstSectionText
 
 - (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
 {
-    if (section == 1)
+    if ([_data[section][0][@"type"] isEqualToString:kCellTypeMap])
```
Comment (PaulStets) [inline Sources/Controllers/2.0/Settings/ImportExport/OAPluginInstalledViewController.mm:363]: Error-prone code. Please use variables for sections and assign them to section indexes if they exist, otherwise leave them at -1.

File: Sources/Plugins/SRTMPlugin/OASRTMPlugin.mm [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+#define PLUGIN_ID kInAppId_Addon_Srtm
+
+@implementation OASRTMPlugin
+
+- (instancetype) init
+{
+    self = [super init];
+    if (self)
+    {
+        [OAApplicationMode regWidgetVisibility:PLUGIN_ID am:nil];
```
Comment (PaulStets) [inline Sources/Plugins/SRTMPlugin/OASRTMPlugin.mm:23]: Cannot find this code in Android

File: Sources/Controllers/2.0/OAMapSettingsContourLinesScreen.mm [hunk 5]
```diff
@@ -365,28 +365,11 @@ - (void) generateData
...
-                }
-                _mapItems = availableItems;
-            }
-            
-            [self generateData];
-            [tblView reloadData];
-        }
+    CLLocationCoordinate2D loca = [OAResourcesUIHelper getMapLocation];
+    [OAResourcesUIHelper getMapsForType:OsmAnd::ResourcesManager::ResourceType::SrtmMapRegion latLon:loca onComplete:^(NSArray<OARepositoryResourceItem *>* res) {
+        _mapItems = res;
```
Comment (PaulStets) [inline Sources/Controllers/2.0/OAMapSettingsContourLinesScreen.mm:370]: Why did you remove the ```@synchronized(_dataLock)``` block? Also, check if this code is always called from the main thread.

File: Sources/Helpers/OAResourcesUIHelper.mm [hunk 6]
```diff
@@ -56,6 +57,12 @@ - (void) updateSize
     // override
 }
 
+- (BOOL) isInstalled
+{
+    NSString *path = [[OsmAndApp.instance.dataPath stringByAppendingPathComponent:@"Resources"] stringByAppendingPathComponent:self.resourceId.toNSString()];
```
Comment (PaulStets) [inline Sources/Helpers/OAResourcesUIHelper.mm:62]: Use existring constant for Resources (OAIndexConstants)

File: Sources/Helpers/OAResourcesUIHelper.mm [hunk 7]
```diff
@@ -188,6 +195,80 @@ + (NSString *) resourceTypeLocalized:(OsmAnd::ResourcesManager::ResourceType)typ
...
+    //MapStyle,
+    //MapStylesPresets,
+    //OnlineTileSources,
+    
+    return OsmAnd::ResourcesManager::ResourceType::Unknown;
+}
+
++ (NSString *) iconNameByResourseType:(OsmAnd::ResourcesManager::ResourceType)type
+{
+    if (type == OsmAnd::ResourcesManager::ResourceType::MapRegion)
```
Comment (PaulStets) [inline Sources/Helpers/OAResourcesUIHelper.mm:240]: The first two conditions are redundant, because @"ic_custom_map" is default.

File: Sources/Helpers/OAResourcesUIHelper.mm [hunk 8]
```diff
@@ -453,105 +534,201 @@ + (OAWorldRegion*) findRegionOrAnySubregionOf:(OAWorldRegion*)region
     return items;
 }
 
-+ (void) requestMapDownloadInfo:(CLLocationCoordinate2D)coordinate resourceType:(OsmAnd::ResourcesManager::ResourceType)resourceType onComplete:(void (^)(NSArray<OAResourceItem *>*))onComplete
++ (NSArray<OAResourceItem *> *) findIndexItemsAt:(NSArray<NSString *> *)names type:(OsmAnd::ResourcesManager::ResourceType)type includeDownloaded:(BOOL)includeDownloaded limit:(NSInteger)limit
 {
-    NSMutableArray<OAResourceItem *>* res;
-    res = [NSMutableArray new];
-    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
```
Comment (PaulStets) [inline Sources/Helpers/OAResourcesUIHelper.mm:460]: Plaese check that you didn't remove the asynchronous behavior of this code.

File: Sources/Controllers/2.0/Settings/ImportExport/OAPluginInstalledViewController.mm [hunk 9]
```diff
@@ -0,0 +1,551 @@
...
+    else
+    {
+        NSMutableParagraphStyle *style = [[NSMutableParagraphStyle alloc] init];
+        [style setLineSpacing:6];
+        attrString = [[NSAttributedString alloc] initWithString:descriptionText attributes:@{NSParagraphStyleAttributeName : style}];
+    }
+    return [OAUtilities setupTableHeaderViewWithText:attrString tintColor:UIColor.whiteColor icon:_plugin.getLogoResource iconFrameSize:48. iconBackgroundColor:UIColorFromRGB(color_primary_purple) iconContentMode:UIViewContentModeScaleAspectFit iconYOffset:48.];
+}
+
+- (NSInteger) getTypeForSection:(NSInteger)section
```
Comment (PaulStets) [inline Sources/Controllers/2.0/Settings/ImportExport/OAPluginInstalledViewController.mm:368]: Create an enum instead

File: Sources/Controllers/2.0/Settings/ImportExport/OAPluginInstalledViewController.mm [hunk 10]
```diff
@@ -0,0 +1,551 @@
...
+    NSString *type = _data[indexPath.section][indexPath.row][@"type"];
+    if ([type isEqualToString:kCellTypeMap])
+        [self onItemClicked:indexPath];
+
+    [self.tableView deselectRowAtIndexPath:indexPath animated:NO];
+}
+
+#pragma mark - Selectors
+
+- (void) accessoryButtonTapped:(UIControl *)button withEvent:(UIEvent *)event
```
Comment (PaulStets) [inline Sources/Controllers/2.0/Settings/ImportExport/OAPluginInstalledViewController.mm:416]: tapped, pressed and clicked in one View controller, please use ```pressed``` only

File: Sources/Controllers/2.0/Settings/ImportExport/OAPluginInstalledViewController.mm [hunk 11]
```diff
@@ -0,0 +1,551 @@
...
+
+- (OARepositoryResourceItem *) getMapItem:(NSIndexPath *)indexPath
+{
+    return (OARepositoryResourceItem *)_suggestedMaps[indexPath.row];
+}
+
+- (void) onItemClicked:(NSIndexPath *)indexPath
+{
+    OARepositoryResourceItem *mapItem = [self getMapItem:indexPath];
+        if (mapItem.downloadTask != nil)
```
Comment (PaulStets) [inline Sources/Controllers/2.0/Settings/ImportExport/OAPluginInstalledViewController.mm:438]: Formatting

File: Sources/Controllers/2.0/Settings/ImportExport/OAPluginInstalledViewController.mm [hunk 12]
```diff
@@ -0,0 +1,551 @@
...
+        return;
+ 
+    dispatch_async(dispatch_get_main_queue(), ^{
+        [self refreshDownloadingContent:task.key];
+    });
+}
+
+- (void) refreshDownloadingContent:(NSString *)downloadTaskKey
+{
+    for (int i = 0; i < _suggestedMaps.count; i++)
```
Comment (PaulStets) [inline Sources/Controllers/2.0/Settings/ImportExport/OAPluginInstalledViewController.mm:511]: Use NSInteger

File: Sources/Helpers/OAResourcesUIHelper.mm [hunk 13]
```diff
@@ -56,6 +57,12 @@ - (void) updateSize
     // override
 }
 
+- (BOOL) isInstalled
```
Comment (PaulStets) [inline Sources/Helpers/OAResourcesUIHelper.mm:60]: You can check if the resource is installed via ```OsmAnd::ResourcesManager::ResourceOrigin::Installed```, so this method is probably incorrect and unnecessary.

File: Sources/Helpers/OASuggestedDownloadsItem.h [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+@interface OASuggestedDownloadsItem : OASettingsItem
+
+@property (nonatomic, readonly) NSString *scopeId;
+@property (nonatomic, readonly) NSString *searchType;
+@property (nonatomic, readonly) NSArray<NSString *> *names;
+@property (nonatomic, readonly) NSInteger limit;
+
+@property (nonatomic, readonly) NSMutableArray *items;
+
+- (OASuggestedDownloadsItem *) initWithScopeId:(NSString *)scopeId searchType:(NSString *)searchType names:(NSArray<NSString *> *)names limit:(NSInteger)limit;
```
Comment (PaulStets) [inline Sources/Helpers/OASuggestedDownloadsItem.h:24]: You should return instancetype here

File: Sources/Controllers/2.0/OAMapSettingsContourLinesScreen.mm [hunk 15]
```diff
@@ -365,25 +365,11 @@ - (void) generateData
...
-                    {
-                        OARepositoryResourceItem *resource = (OARepositoryResourceItem*)item;
-                        [availableItems addObject:resource];
-                    }
-                }
-                _mapItems = availableItems;
-            }
-            
+            _mapItems = res;
             [self generateData];
```
[the remark this answers is not in the document]
Comment (nnngrach) [reply Sources/Controllers/2.0/OAMapSettingsContourLinesScreen.mm:373]: https://github.com/osmandapp/OsmAnd-iOS/blob/8597735b3f87ca15a7d4713b0fc26e034e012c6d/Sources/Helpers/OAResourcesUIHelper.mm#L712

File: Sources/Controllers/2.0/Settings/ImportExport/OAPluginInstalledViewController.mm (hunk 9 above)
Comment (nnngrach) [reply Sources/Controllers/2.0/Settings/ImportExport/OAPluginInstalledViewController.mm:370]: https://github.com/osmandapp/OsmAnd-iOS/pull/1330/files#diff-9e2c29bfedfb179a219277b1e0b1fa32a5ea41df01bd2b6e81ca3395dd9420dcR36
